### PR TITLE
#5125 RevisionGrid Graph: Branches and tags in tooltip.

### LINKED ISSUE #4450: Commit branch name in commits list
**Do you want to request a *feature* or report a *bug*?**
feature

**What is the current behavior?**
"Revision graph" doesn't show hint in which branch we have "Dot" (commit). Commits list also doesn't show this information. User have to use Details panel / Commit tab, or double click on commit.

**What is the expected behavior?**
One or few from next list:
- On mouse hover on graph line, output hint about branch name
- On mouse hover on graph commit dot, output hint about commit hash AND branch name
- Possibility to turn on commits list column where user can see branch name

**Environment you encounter the issue:**

- GitExtensions version: 2.51
- GIT version: 2.16.1
- OS version: Windows 10 1709 b16299.214

**Did this work in previous version of GitExtensions (which)?**
no

### LINKED ISSUE #5125: RevisionGrid Graph: Show more informative tooltips
**Do you want to request a *feature* or report a *bug*?**

feature

**What is the current behavior?**

tooltip with commit hash of wrong lane
grafik

**What is the expected behavior?**

commit hash, [*later:* branches contained in / being at] and commit heading of lane under mouse
grafik
grafik
grafik

**Environment you encounter the issue:**

- GitExtensions version: v3-ux, master
- GIT version: N/A (2.17.windows.2)
- OS version: N/A (Windows 7)
- .NET version: N/A

**Did this work in previous version of GitExtensions (which)?**

PR #5017 but with high load by firing git.exe for display of branch info

### PR DESCRIPTION
**Pull request by mstv**
Fixes #5125 
Fixes #4450

#### Changes proposed in this pull request:
- extend tooltips for the Revision Graph with
  - commit hash for all lanes (prefixed with `*` if it is the lane of the commit in the current row)
  - branch to which was committed (as far as possible to determine)
  - branches and tags the commit is contained in
- limit the number of lines in tooltips to 30
- insert litte spaces into the commit hash
 
#### Screenshots before and after (if PR changes UI):
- before:
grafik
grafik
grafik
- after:
grafik
grafik
grafik
- before: too long tooltip flashes for short time
- after:
grafik
- after: branch committed to
grafik
grafik
grafik
- before: no tooltips on lines with artificial commits not even for other lanes
- after:
grafik
grafik
grafik

#### What did I do to test the code and ensure quality:
- compared branches and tags contained in with the commit info of many revisions
- checked with artificial commits and stash
- created test repository with non-standard merge commit messages

#### Has been tested on (remove any that don't apply):
- GIT: 2.18.0.w.x64
- Windows 7

### CODE REVIEW

File: GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs [hunk 1]
```diff
@@ -66,8 +67,26 @@ public override bool TryGetToolTip(DataGridViewCellMouseEventArgs e, GitRevision
...
 
-            toolTip = revision.Guid;
+            toolTip = Format(revision.Guid);
             return true;
         }
+
+        public static string Format(string guid)
+        {
+            var text = new StringBuilder();
+            for (int pos = 0, length = guid.Length; pos < length; pos += 4)
```
Comment (gerhardol) [inline GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs]: why not 
for (int pos = 0; pos < guid.Length; pos += 4)

Comment (mstv) [reply GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs]: I do not know and shall not know the implementation of the *property* string.Length. It might add some overhead getting it multiple times.
I am afraid the compiler/optimizer might not recognize that `guid` and its length is constant. (C# has only a limited concept of constant/read-only expressions.)
Code for a single tooltip is not time-critical. Though I prefer to always consider the performance.
But if you tell me that it is not necessary (at least for string), I will use the shorter implementation and will omit such helper variables.

Comment (gerhardol) [reply GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs]: comma operator is a little obscure (I thought it was not even allowed in C# for())
I do not know how the compiler optimizes this, but would not consider this critical anyway and prefer readability. Not a strong opinion.

Comment (RussKie) [reply GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs]: I think you're trying to optimise prematurely here. .NET strings are immutable, so a string's length is a constant.

Slight OT, most of the heavy optimisation lifting these days is done by the JIT, and it is pretty smart.
This article may be useful: http://benhall.io/c-debug-vs-release-builds-and-debugging-in-visual-studio-from-novice-to-expert-in-one-blog-article/

File: GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs [hunk 2]
```diff
@@ -66,8 +67,26 @@ public override bool TryGetToolTip(DataGridViewCellMouseEventArgs e, GitRevision
                 return false;
             }
 
-            toolTip = revision.Guid;
+            toolTip = Format(revision.Guid);
             return true;
         }
+
+        public static string Format(string guid)
```
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs]: After staring at this method for a bit I understand it formats the guid into four digit chunks separated by a space.
What escapes me is why it is needed. We don't format guids anywhere else in the app. Nor I haven't seen it is done anywhere else in relation to git repos.

No offence, but to me this method adds unnecessary complexity. It also negatively affects the performance as it will be run everytime and result in memory allocations.

Comment (vbjay) [reply GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs]: https://docs.microsoft.com/en-us/dotnet/api/system.guid.tostring?view=netframework-4.7.2#System_Guid_ToString_System_String_

Comment (RussKie) [reply GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs]: Yes, that's too. though because it is a string not a guid, we'll have to cast, that will lead to additional allocations.
In this case I don't believe it is necessary at all.

Comment (mstv) [reply GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs]: This method is used only when a tooltip is to be displayed. This should have no substancial effect on the performance.

I cannot read such blobs of hexadecimal digits easily. It is a constant annoyance to me if they are longer than 6-7 digits.
To give some optical guiding, I inserted little gaps (about half a space).
You can see it on the screenshots. Give it a try, please!

A Git GUID is not a System.GUID (40 vs. 32 bits), @vbjay.

Comment (vbjay) [reply GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs]: My point was to make the git guid format correctly.  Define the formatting in the class.  You can even use format strings like System.GUID does. Then it will be as simple as something like this 

```c#
theGUID.ToString("D"); 
```

File: GitUI/UserControls/RevisionGrid/Columns/GraphColumnProvider.cs [hunk 3]
```diff
@@ -556,9 +552,32 @@ string GetLaneInfo(int x, int rowIndex)
...
+                                    if (references.MergedWith.IsNotNullOrWhitespace())
+                                    {
+                                        laneInfoText.AppendFormat(" (merged with {0})", references.MergedWith);
+                                    }
+                                }
+
+                                laneInfoText.Append(Format("\nContained in branches: {0}\n", references.Branches));
+                                laneInfoText.Append(Format("\nContained in tags: {0}\n", references.Tags));
+
+                                string Format<T>(string format, IEnumerable<T> list)
```
Comment (vbjay) [inline GitUI/UserControls/RevisionGrid/Columns/GraphColumnProvider.cs]: This doesn't belong here.  We have extension methods that can be used or modified.

Comment (vbjay) [reply GitUI/UserControls/RevisionGrid/Columns/GraphColumnProvider.cs]: I would move it to there or even a list extensions class and call it ToCSV or something like it.

**Comment by RussKie** [comment]
With this change i get one of the repos at work locking up and becoming
unresponsive for seconds...
Tooltips also cover half the screen, which is more an annoyance than help.

File: GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs (hunk 2 above)
Comment (RussKie) [reply GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs]: > I cannot read such blobs of hexadecimal digits easily. It is a constant annoyance to me if they are longer than 6-7 digits.

Whilst I can empathise your feeling I can not justify the inconsistency in UX this change introduces. Hashes are universally rendered as one unbroken string - in GE, on GH, and other apps I've seen.
We had a discussion (#4420) whether we should dynamically calculate hash sizes, but in the end it was concluded to be a perf overhead.

**Comment by RussKie** [comment]
I support the idea of tooltips providing meaningful information for hovered commits. Certainly there are few gotchas and traps along this path.

IIRC for perf reasons we do not display the full body of a commit until it has been loaded. 
A lazy load on hover is probably ok, however it needs to be done leveraging `CancellationTokenSequence`, if there is a body - show it, if not - load it, unless the user has moved the mouse and requested another commit body load.

@gerhardol you wrote in https://github.com/gitextensions/gitextensions/issues/5125#issuecomment-401959907:
> When hovering the branches, I expect to see branch information, probably tags too. (not like this in v3-ux branch)

I'm also finding the proposed amount of information overbearing. Why is it important to you to see _all_ branches and tags?
If I really want to know what tags or branches (there may be dozens of those in large repos) a particular commit belongs to, or to see full information about the commit - I would click on it. 
Here's a Michael's example:
grafik
In some of my cases, I get dozens of branches spanning 2 or 3 lines across the screen.


Tooltips are supposed to be lightweight and instantaneous; in my case these tooltips render some of the repos I work on irresponsive and sluggish.

**Comment by gerhardol** [comment]
Repeating https://github.com/gitextensions/gitextensions/issues/5125#issuecomment-401959907
> When hovering the commit headline I expect to get a tooltip also when there is a one line message (especially when branch tags are replacing the message).

I still would like to see that here, even if tooltips are hidden only when all of the message is not visible

> I have removed them from the tooltip again. 

> New suggestion: Why don't we display all branches and tags in the Commit Info? There is a scrollbar available in contrast to tooltips.

More

**Comment by RussKie** [comment]
> New suggestion: Why don't we display all branches and tags in the Commit Info? There is a scrollbar available in contrast to tooltips.

Yes, I have tinkered with the commit info awhile back, but it is a mere link in a long chain of things that need fixing... 😞 
There is an intent to give the commit info panel an overhaul once v3.00 is shipped.

**Comment by RussKie** [comment]
Hi Michael, I took a liberty of refactoring your code to make it more modular and testable. I also added a test scaffold and added few tests, though obviously more tests are needed.
As usual eager to hear your thoughts.

**Comment by gerhardol** [comment]
@mstv What is the status for this PR?
When playing around with the linux repo, it is obvious that the size of repos must be limited and there must be a delay before the tooltip is shown, there is a lot of info moving the mouse over the grid.

**Comment by gerhardol** [comment]
@mstv What is the status for this PR?

**Comment by mstv** [comment]
I was very busy in RL. I have just resolved the merge conflicts and restored a minor functionality.
I am cloning the Linux repo now.
I will add test cases.

**Comment by RussKie** [comment]
> Why does AppVeyor return a different NUnit test result than the local
execution?

what do you mean?

**Comment by mstv** [comment]
The AppVeyor build succeeded in the meantime. I did not notice that it was still running and that the link still pointed to the previous build report.

**Comment by gerhardol** [comment]
I saw some performance issues in Linux, but that depends a lot of the commits you happen to play with.
This branch seem better to me now, as it limits the size of tooltips. 
Still approving this PR
No review of testcases

File: GitUI/UserControls/RevisionGrid/Graph/LaneInfoProvider.cs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+using System.Text;
+using System.Text.RegularExpressions;
+using JetBrains.Annotations;
+using ResourceManager;
+
+namespace GitUI.UserControls.RevisionGrid.Graph
+{
+    internal sealed class LaneInfoProvider
+    {
+        private readonly TranslationString _NoInfoText = new TranslationString("sorry, this commit seems to be not loaded");
```
Comment (drewnoakes) [inline GitUI/UserControls/RevisionGrid/Graph/LaneInfoProvider.cs]: Nit: Capitalisation

File: GitUI/UserControls/RevisionGrid/Graph/LaneNodeLocator.cs [hunk 5]
```diff
@@ -22,40 +22,42 @@ public LaneNodeLocator(ILaneRowProvider laneRowProvider)
...
                     }
                 }
-
-                return (null, false);
             }
+
+            return (null, false);
         }
     }
 }
```
Comment (illfated) [inline GitUI/UserControls/RevisionGrid/Graph/LaneNodeLocator.cs]: Nitpick: missing NewLine at EOF

File: UnitTests/GitUITests/UserControls/RevisionGrid/Graph/LaneNodeLocatorTests.cs [hunk 6]
```diff
@@ -39,19 +39,40 @@ public void FindPrevNode_should_return_null_if_laneWidth_less_than_one()
...
             laneRow.Count.Returns(10);
             _laneRowProvider.GetLaneRow(100).Returns(x => laneRow);
 
-            // lane >= _laneRowProvider.GetLaneRow(rowIndex).Count  TODO: rowIndex cannot be compared to laneRow.Count
-            _laneNodeLocator.FindPrevNode(0, 100, 1).Should().Be((null, false));
+            // lane >= _laneRowProvider.GetLaneRow(rowIndex).Count
+            _laneNodeLocator.FindPrevNode(10, 100, 1).Should().Be((null, false));
         }
     }
 }
```
Comment (illfated) [inline UnitTests/GitUITests/UserControls/RevisionGrid/Graph/LaneNodeLocatorTests.cs]: Nitpick: missing NewLine at EOF

File: GitUI/UserControls/RevisionGrid/Graph/LaneNodeLocator.cs (hunk 5 above)
Comment (mstv) [reply GitUI/UserControls/RevisionGrid/Graph/LaneNodeLocator.cs]: @RussKie, older source files end with a NewLine, newer ones do not. What is the preferred style?

File: GitUI/UserControls/RevisionGrid/Graph/LaneInfoProvider.cs [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+            /// <param name="descJunction">
+            /// the descending junction the node is part of
+            /// (used for the decision whether the node belongs the first or second branch of the merge)
+            /// </param>
+            private bool CheckForMerge([NotNull] Node node, [CanBeNull] Junction descJunction)
+            {
+                var isTheFirstBranch = descJunction == null || node.Ancestors.Count == 0 || node.Ancestors.First() == descJunction;
+                string mergedInto;
+                string mergedWith;
+                (mergedInto, mergedWith) = ParseMergeMessage(node, isTheFirstBranch);
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/Graph/LaneInfoProvider.cs]: @RussKie, why did you remove `appendPullRequest: ` from the argument `isTheFirstBranch` given in the call of `ParseMergeMessage`?

File: GitUI/UserControls/RevisionGrid/Graph/LaneInfoProvider.cs [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+            /// First/second branch does not matter because it is the message of the current node.
+            /// </summary>
+            /// <param name="node">the node of the revision to evaluate</param>
+            /// <param name="descJunction">
+            /// the descending junction the node is part of
+            /// (used for the decision whether the node belongs the first or second branch of the merge)
+            /// </param>
+            private bool CheckForMerge([NotNull] Node node, [CanBeNull] Junction descJunction)
+            {
+                var isTheFirstBranch = descJunction == null || node.Ancestors.Count == 0 || node.Ancestors.First() == descJunction;
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/Graph/LaneInfoProvider.cs]: @RussKie, regarding `var`: Yes, often, the type can easily be deduced from the variable's identifier.
But in case of `bool`, I must think a little.
`int` vs. `uint` is not obvious.
Especially when reading diffs without Intellisense, I prefer explicit types to `var`.

What does your inner Style Cop say? "Use `var` whenever possible"?
For shortness of the code? Or for simplicity of the style rule?

Comment (RussKie) [reply GitUI/UserControls/RevisionGrid/Graph/LaneInfoProvider.cs]: It would be hypocritical of me to tell contributors use one way or the other as I'm a little inconsistent myself - if a type's name is short (e.g. `int`, `string`, `bool`) I type it; for complex types I use `var` unless the variable of an interface type. 
Sometimes I let Re# cleaner out of its cage and let it reformat the code as it sees fitting.

File: GitUI/UserControls/RevisionGrid/Graph/LaneInfoProvider.cs (hunk 7 above)
Comment (RussKie) [reply GitUI/UserControls/RevisionGrid/Graph/LaneInfoProvider.cs]: I don't remember tbh, I suspect it may have been Re# cleaning the code.

**Comment by mstv** [comment]
Maybe, CodeFactor should ignore NUnit tests.

[Codecov details](https://codecov.io/gh/gitextensions/gitextensions/tree/ec9a44acfbf3e1fca254b3cc6bb9e764e604f641/GitUI/UserControls/RevisionGrid/Graph)

**Comment by RussKie** [comment]
I'm really sorry @mstv, but it looks like you will have to re-do some of this work, as this PR can't go before #5610. Henk has rewritten the graph, that saw some of the files changed in this PR completely changed or removed...

**Comment by spdr870** [comment]
> I'm really sorry @mstv, but it looks like you will have to re-do some of this work, as this PR can't go before #5610. Henk has rewritten the graph, that saw some of the files changed in this PR completely changed or removed...

Sorry..:( On the positive side; getting the refs from all children from a given commit is a lot easier not, and doesn't require the graph to be loaded! Its also a lot quicker. If you want, I could write the method to do this.

**Comment by mstv** [comment]
> getting the refs from all children from a given commit is a lot easier now, and doesn't require the graph to be loaded! Its also a lot quicker. If you want, I could write the method to do this.

We tried to show all refs in the tooltip. It was too much information and covered too much of the window. Instead, all refs should be displayed in the commit info because scrollbars are available there.

This PR deals only with the tooltips for branch lanes in the graph. It shouldn't be too difficult to adapt it to the new data structure because it must contain the same essential information. The basic algorithm should still work.

**Comment by mstv** [comment]
Superseded by #5643.
